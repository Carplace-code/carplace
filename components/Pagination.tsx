"use client";

import {
  Pagination,
  PaginationContent,
  PaginationEllipsis,
  PaginationItem,
  PaginationLink,
  PaginationNext,
  PaginationPrevious,
} from "@/components/ui/pagination";

interface PaginationProps {
  page: number;
  pageCount: number;
  onPageChange: (page: number) => void;
  disabled?: boolean;
}

export default function PaginationComponent({ page, pageCount, onPageChange, disabled = false }: PaginationProps) {
  const getVisiblePages = () => {
    const delta = 2;
    const start = Math.max(1, page - delta);
    const end = Math.min(pageCount, page + delta);
    return Array.from({ length: end - start + 1 }, (_, i) => start + i);
  };

  const visiblePages = getVisiblePages();
  const showStartEllipsis = visiblePages[0] > 2;
  const showEndEllipsis = visiblePages[visiblePages.length - 1] < pageCount - 1;

  return (
    <div className="flex flex-col items-center gap-4">
      <Pagination>
        <PaginationContent>
          {/* Previous button */}
          <PaginationItem>
            <PaginationPrevious
              onClick={() => !disabled && page > 1 && onPageChange(page - 1)}
              className={`${
                page === 1 || disabled ? "pointer-events-none opacity-50" : "hover:bg-accent cursor-pointer"
              }`}
            />
          </PaginationItem>

          {/* First page */}
          {visiblePages[0] > 1 && (
            <PaginationItem>
              <PaginationLink
                onClick={() => !disabled && onPageChange(1)}
                className={disabled ? "pointer-events-none opacity-50" : "cursor-pointer"}
              >
                1
              </PaginationLink>
            </PaginationItem>
          )}

          {/* Start ellipsis */}
          {showStartEllipsis && (
            <PaginationItem>
              <PaginationEllipsis />
            </PaginationItem>
          )}

          {/* Visible page numbers */}
          {visiblePages.map((p) => (
            <PaginationItem key={p}>
              <PaginationLink
                onClick={() => !disabled && onPageChange(p)}
                isActive={p === page}
                className={disabled ? "pointer-events-none opacity-50" : "cursor-pointer"}
              >
                {p}
              </PaginationLink>
            </PaginationItem>
          ))}

          {/* End ellipsis */}
          {showEndEllipsis && (
            <PaginationItem>
              <PaginationEllipsis />
            </PaginationItem>
          )}

          {/* Last page */}
          {visiblePages[visiblePages.length - 1] < pageCount && (
            <PaginationItem>
              <PaginationLink
                onClick={() => !disabled && onPageChange(pageCount)}
                className={disabled ? "pointer-events-none opacity-50" : "cursor-pointer"}
              >
                {pageCount}
              </PaginationLink>
            </PaginationItem>
          )}

          {/* Next button */}
          <PaginationItem>
            <PaginationNext
              onClick={() => !disabled && page < pageCount && onPageChange(page + 1)}
              className={`${
                page === pageCount || disabled ? "pointer-events-none opacity-50" : "hover:bg-accent cursor-pointer"
              }`}
            />
          </PaginationItem>
        </PaginationContent>
      </Pagination>

      {/* Page info */}
      <div className="text-muted-foreground text-sm">
        PÃ¡gina {page} de {pageCount}
        {disabled && <span className="ml-2">(Cargando...)</span>}
      </div>
    </div>
  );
}
