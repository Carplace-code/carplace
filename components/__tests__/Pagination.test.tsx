// components/__tests__/Pagination.test.tsx
import { render } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import Pagination from "../Pagination";

describe("Pagination", () => {
  it("renders correctly and matches snapshot", () => {
    const onPageChange = vi.fn();
    const { container } = render(<Pagination page={2} pageCount={5} onPageChange={onPageChange} />);
    expect(container).toMatchSnapshot();
  });
});
