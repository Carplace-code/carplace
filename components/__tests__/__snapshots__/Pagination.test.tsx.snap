// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col items-center gap-4"
  >
    <nav
      aria-label="pagination"
      class="mx-auto flex w-full justify-center"
      data-slot="pagination"
      role="navigation"
    >
      <ul
        class="flex flex-row items-center gap-1"
        data-slot="pagination-content"
      >
        <li
          data-slot="pagination-item"
        >
          <a
            aria-label="Go to previous page"
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:text-accent-foreground dark:hover:bg-accent/50 h-9 py-2 has-[>svg]:px-3 gap-1 px-2.5 sm:pl-2.5 hover:bg-accent cursor-pointer"
            data-slot="pagination-link"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-chevron-left"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m15 18-6-6 6-6"
              />
            </svg>
            <span
              class="hidden sm:block"
            >
              Previous
            </span>
          </a>
        </li>
        <li
          data-slot="pagination-item"
        >
          <a
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 size-9 cursor-pointer"
            data-active="false"
            data-slot="pagination-link"
          >
            1
          </a>
        </li>
        <li
          data-slot="pagination-item"
        >
          <a
            aria-current="page"
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:bg-accent hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 size-9 cursor-pointer"
            data-active="true"
            data-slot="pagination-link"
          >
            2
          </a>
        </li>
        <li
          data-slot="pagination-item"
        >
          <a
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 size-9 cursor-pointer"
            data-active="false"
            data-slot="pagination-link"
          >
            3
          </a>
        </li>
        <li
          data-slot="pagination-item"
        >
          <a
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 size-9 cursor-pointer"
            data-active="false"
            data-slot="pagination-link"
          >
            4
          </a>
        </li>
        <li
          data-slot="pagination-item"
        >
          <a
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 size-9 cursor-pointer"
            data-slot="pagination-link"
          >
            5
          </a>
        </li>
        <li
          data-slot="pagination-item"
        >
          <a
            aria-label="Go to next page"
            class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:text-accent-foreground dark:hover:bg-accent/50 h-9 py-2 has-[>svg]:px-3 gap-1 px-2.5 sm:pr-2.5 hover:bg-accent cursor-pointer"
            data-slot="pagination-link"
          >
            <span
              class="hidden sm:block"
            >
              Next
            </span>
            <svg
              aria-hidden="true"
              class="lucide lucide-chevron-right"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m9 18 6-6-6-6"
              />
            </svg>
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="text-muted-foreground text-sm"
    >
      PÃ¡gina 
      2
       de 
      5
    </div>
  </div>
</div>
`;
